float32[] t
float32[] x
float32[] y